File: scripts/apply-cleanup.js
#!/usr/bin/env node
// Removes confirmed-stale slugs (category=stale) from app/companies/*, and
// optionally appends re-homed slugs from a portal-discovery report.
// Dry-run by default. Pass --apply to actually mutate files.
//
//   node scripts/apply-cleanup.js                                              # today's stale report, dry-run
//   node scripts/apply-cleanup.js --apply                                      # remove stale
//   node scripts/apply-cleanup.js --apply --rehome reports/portal-discovery-YYYY-MM-DD.csv
//   node scripts/apply-cleanup.js --date 2026-05-17
//
// Git is the audit trail — review `git diff` after applying.

import { readdirSync, readFileSync, writeFileSync, existsSync } from 'fs';

const PORTAL_DIRS = {
    greenhouse: { dir: 'app/companies/greenhouse', format: 'csv', addFile: 'gh-io.csv' },
    lever:      { dir: 'app/companies/lever',      format: 'csv', addFile: 'lever.csv' },
    ashby:      { dir: 'app/companies/ashbyhq',    format: 'csv', addFile: 'ash.csv' },
    oracloud:   { dir: 'app/companies/oracloud',   format: 'json', slugField: 'companyName' },
    workday:    { dir: 'app/companies/workday',    format: 'json', slugField: 'name' },
};

const argv = (flag) => {
    const i = process.argv.indexOf(flag);
    return i >= 0 && process.argv[i + 1] ? process.argv[i + 1] : null;
};

const args       = process.argv.slice(2);
const apply      = args.includes('--apply');
const date       = argv('--date') || new Date().toISOString().slice(0, 10);
const rehomePath = argv('--rehome');

const reportPath = `reports/stale-companies-${date}.csv`;
if (!existsSync(reportPath)) {
    console.error(`Report not found: ${reportPath}`);
    process.exit(1);
}

// Parse report → { [portal]: Set<slug> } (stale only, never unknown).
const staleByPortal = {};
const lines = readFileSync(reportPath, 'utf8').split('\n').slice(1);
for (const line of lines) {
    if (!line.trim()) continue;
    const fields = line.split(',');
    if (fields.length < 3) {
        console.warn(`Skipping malformed stale row: ${line}`);
        continue;
    }
    const [category, portal, slug] = fields;
    if (category !== 'stale' || !portal || !slug) continue;
    (staleByPortal[portal] ??= new Set()).add(slug.toLowerCase());
}

const sourceFiles = (dir, format) => readdirSync(dir)
    .filter((f) => f.endsWith(`.${format}`) && !f.includes('test') && !f.endsWith('.bak'));

let totalRemoved = 0;
const summary    = [];

for (const [portal, staleSlugs] of Object.entries(staleByPortal)) {
    const cfg = PORTAL_DIRS[portal];
    if (!cfg) { console.warn(`Skipping unknown portal: ${portal}`); continue; }

    for (const file of sourceFiles(cfg.dir, cfg.format)) {
        const path = `${cfg.dir}/${file}`;
        const raw  = readFileSync(path, 'utf8');

        let kept, removed;
        if (cfg.format === 'csv') {
            const all = raw.split('\n');
            removed   = all.filter((l) => staleSlugs.has(l.trim().toLowerCase())).length;
            kept      = all.filter((l) => !staleSlugs.has(l.trim().toLowerCase()));
        } else {
            const all  = JSON.parse(raw);
            const next = all.filter((r) => !staleSlugs.has(String(r?.[cfg.slugField] ?? '').toLowerCase()));
            removed    = all.length - next.length;
            kept       = next;
        }

        if (removed === 0) continue;

        totalRemoved += removed;
        summary.push(`  ${portal}/${file}: ${removed} removed`);

        if (apply) {
            const out = cfg.format === 'csv' ? kept.join('\n') : JSON.stringify(kept, null, 2) + '\n';
            writeFileSync(path, out);
        }
    }
}

console.log(`Stale report: ${reportPath}`);
console.log(`Mode:         ${apply ? 'APPLY (files written)' : 'dry-run'}`);
if (summary.length === 0) {
    console.log('No matching slugs found in source files.');
} else {
    console.log(summary.join('\n'));
    console.log(`Total removed: ${totalRemoved} ${apply ? '' : '(would be)'}.`);
}

// ─── Re-home phase ────────────────────────────────────────────────────────────
if (rehomePath) {
    if (!existsSync(rehomePath)) {
        console.error(`\nDiscovery report not found: ${rehomePath}`);
        process.exit(1);
    }

    // discovery columns: slug,original_portal,found_in,found_url,status
    const addByPortal = {};
    const lines = readFileSync(rehomePath, 'utf8').split('\n').slice(1);
    for (const line of lines) {
        if (!line.trim()) continue;
        const fields = line.split(',');
        if (fields.length < 3) {
            console.warn(`Skipping malformed rehome row: ${line}`);
            continue;
        }
        const [slug, , found_in] = fields;
        if (!slug || !found_in || found_in === 'none') continue;
        (addByPortal[found_in] ??= new Set()).add(slug.toLowerCase());
    }

    console.log('\n--- Re-home ---');
    console.log(`Discovery report: ${rehomePath}`);

    let totalAdded = 0;
    const addSummary = [];

    for (const [portal, slugs] of Object.entries(addByPortal)) {
        const cfg = PORTAL_DIRS[portal];
        if (!cfg || cfg.format !== 'csv' || !cfg.addFile) {
            console.warn(`Skipping re-home for ${portal} (no CSV target)`);
            continue;
        }
        const path = `${cfg.dir}/${cfg.addFile}`;
        const raw  = existsSync(path) ? readFileSync(path, 'utf8') : '';
        const existing = new Set(raw.split('\n').map((l) => l.trim().toLowerCase()).filter(Boolean));

        const toAdd = [...slugs].filter((s) => !existing.has(s)).sort();
        if (toAdd.length === 0) continue;

        totalAdded += toAdd.length;
        addSummary.push(`  ${portal}/${cfg.addFile}: +${toAdd.length}`);

        if (apply) {
            const sep = raw.length > 0 && !raw.endsWith('\n') ? '\n' : '';
            writeFileSync(path, raw + sep + toAdd.join('\n') + '\n');
        }
    }

    if (addSummary.length === 0) {
        console.log('No new slugs to add (all already present in target CSVs).');
    } else {
        console.log(addSummary.join('\n'));
        console.log(`Total added: ${totalAdded} ${apply ? '' : '(would be)'}.`);
    }
}

if (!apply && (totalRemoved > 0 || rehomePath)) {
    console.log('\nRun again with --apply to write changes.');
}

File: scripts/apply_cleanup.py
#!/usr/bin/env python3
# Removes confirmed-stale slugs (category=stale) from app/companies/*, and
# optionally appends re-homed slugs from a portal-discovery report.
# Dry-run by default. Pass --apply to actually mutate files.
#
#   python scripts/apply_cleanup.py                                              # today's stale report, dry-run
#   python scripts/apply_cleanup.py --apply                                      # remove stale
#   python scripts/apply_cleanup.py --apply --rehome reports/portal-discovery-YYYY-MM-DD.csv
#   python scripts/apply_cleanup.py --date 2026-05-17
#
# Git is the audit trail — review `git diff` after applying.

import argparse
import json
import os
import sys
from datetime import date

PORTAL_DIRS = {
    'greenhouse': {'file': 'app/companies/greenhouse.csv', 'format': 'csv'},
    'lever':      {'file': 'app/companies/lever.csv',      'format': 'csv'},
    'ashby':      {'file': 'app/companies/ashby.csv',      'format': 'csv'},
    'oracloud':   {'file': 'app/companies/oracloud.json',  'format': 'json', 'slug_field': 'companyName'},
    'workday':    {'file': 'app/companies/workday.json',   'format': 'json', 'slug_field': 'name'},
}

parser = argparse.ArgumentParser()
parser.add_argument('--apply',  action='store_true')
parser.add_argument('--date',   default=date.today().isoformat())
parser.add_argument('--rehome', default=None)
args = parser.parse_args()

report_path = f'reports/stale-companies-{args.date}.csv'
if not os.path.exists(report_path):
    print(f'Report not found: {report_path}', file=sys.stderr)
    sys.exit(1)

# Parse report → {portal: set(slug)} (stale only)
stale_by_portal = {}
with open(report_path) as f:
    for line in f.readlines()[1:]:
        line = line.strip()
        if not line:
            continue
        fields = line.split(',')
        if len(fields) < 3:
            print(f'Skipping malformed stale row: {line}', file=sys.stderr)
            continue
        category, portal, slug = fields[0], fields[1], fields[2]
        if category != 'stale' or not portal or not slug:
            continue
        stale_by_portal.setdefault(portal, set()).add(slug.lower())

total_removed = 0
summary = []

for portal, stale_slugs in stale_by_portal.items():
    cfg = PORTAL_DIRS.get(portal)
    if not cfg:
        print(f'Skipping unknown portal: {portal}', file=sys.stderr)
        continue

    path = cfg['file']
    with open(path) as f:
        raw = f.read()

    if cfg['format'] == 'csv':
        all_lines = raw.split('\n')
        kept    = [l for l in all_lines if l.strip().lower() not in stale_slugs]
        removed = len(all_lines) - len(kept)
    else:
        all_entries = json.loads(raw)
        field = cfg['slug_field']
        kept    = [r for r in all_entries if str(r.get(field, '')).lower() not in stale_slugs]
        removed = len(all_entries) - len(kept)

    if removed == 0:
        continue

    total_removed += removed
    summary.append(f'  {path}: {removed} removed')

    if args.apply:
        if cfg['format'] == 'csv':
            out = '\n'.join(kept)
        else:
            out = json.dumps(kept, indent=2) + '\n'
        with open(path, 'w') as f:
            f.write(out)

print(f'Stale report: {report_path}')
print(f'Mode:         {"APPLY (files written)" if args.apply else "dry-run"}')
if not summary:
    print('No matching slugs found in source files.')
else:
    print('\n'.join(summary))
    suffix = '' if args.apply else ' (would be)'
    print(f'Total removed: {total_removed}{suffix}.')

# ── Re-home phase ──────────────────────────────────────────────────────────────
if args.rehome:
    if not os.path.exists(args.rehome):
        print(f'\nDiscovery report not found: {args.rehome}', file=sys.stderr)
        sys.exit(1)

    # discovery columns: slug,original_portal,found_in,found_url,status
    add_by_portal = {}
    with open(args.rehome) as f:
        for line in f.readlines()[1:]:
            line = line.strip()
            if not line:
                continue
            fields = line.split(',')
            if len(fields) < 3:
                print(f'Skipping malformed rehome row: {line}', file=sys.stderr)
                continue
            slug, _, found_in = fields[0], fields[1], fields[2]
            if not slug or not found_in or found_in == 'none':
                continue
            add_by_portal.setdefault(found_in, set()).add(slug.lower())

    print('\n--- Re-home ---')
    print(f'Discovery report: {args.rehome}')

    total_added = 0
    add_summary = []

    for portal, slugs in add_by_portal.items():
        cfg = PORTAL_DIRS.get(portal)
        if not cfg or cfg['format'] != 'csv':
            print(f'Skipping re-home for {portal} (no CSV target)', file=sys.stderr)
            continue

        path = cfg['file']
        raw  = open(path).read() if os.path.exists(path) else ''
        existing = {l.strip().lower() for l in raw.split('\n') if l.strip()}

        to_add = sorted(s for s in slugs if s not in existing)
        if not to_add:
            continue

        total_added += len(to_add)
        add_summary.append(f'  {path}: +{len(to_add)}')

        if args.apply:
            sep = '' if raw.endswith('\n') else '\n'
            with open(path, 'a') as f:
                f.write(sep + '\n'.join(to_add) + '\n')

    if not add_summary:
        print('No new slugs to add (all already present in target CSVs).')
    else:
        print('\n'.join(add_summary))
        suffix = '' if args.apply else ' (would be)'
        print(f'Total added: {total_added}{suffix}.')

if not args.apply and (total_removed > 0 or args.rehome):
    print('\nRun again with --apply to write changes.')
